scipy/special/sf_error_state.cpp

```diff
@@ -1,10 +1,10 @@
 #include <stdlib.h>
-
 #include "sf_error_state.h"
 
 
+
 /* If this isn't volatile clang tries to optimize it away */
-static volatile sf_action_t sf_error_actions[] = {
+static volatile thread_local sf_action_t sf_error_actions[] = {
```

Comment:
does this still need to be volatile if it's thread_local?

Reply:
heh, LLVM was trying to tell the author of the comment above that the code was doing UB...

Reply:
Would it still optimize it away when using both `static` and `thread_local`?

Reply:
No, the comment is alluding to the fact that LLVM was optimizing away a `static volatile` variable storing global state. LLVM was doing this because marking a variable as `volatile` is *not* sufficient to prevent data races, so LLVM detected UB there and optimized it away, because by definition UB can't happen.

If it's `static thread_local` then there's no multithreaded data race and no UB.

Reply:
I didn't write the comment above, but I did copy it from its original home into the `sf_error_state.c`. I don't think the variable being optimized away necessarily had anything to do with data races.  I think what was going on is that `clang` decided this wasn't being used in the translation unit because `scipy_sf_error_set_action` and `scipy_sf_error_get_action` are only called through the Python runtime, and Python's dynamic function lookup makes it tricky to determine that they are actually used through static analysis. A more principled solution would have been to conditionally compile in a `__attribute__((used))` on platforms that support it. If it's working now without the `volatile`, it may just be that Clang's static analyzer has improved.

Reply:
Sorry, my attempted explanation assumed there's some kind of link time optimization going on, but that might not necessarily be the case. I still don't think the original issue had anything to do with a data race, I think the way that these could only be accessed from Python with the GIL in place would have prevented that possibility.